src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/CustomTagEntryFormalizerTests.cs

```diff
@@ -0,0 +1,92 @@
+            Assert.Equal(entry.Level, formalized.Level);
+        }
+
+        [InlineData(null)]
+        [InlineData("")]
+        [Theory]
+
+        public void GivenStandardTagWithoutVR_WhenFormalizing_ThenVRShouldBeFilled(string vr)
+        {
+            DicomTag tag = DicomTag.DeviceSerialNumber;
+            CustomTagEntry entry = new CustomTagEntry(tag.GetPath(), vr, CustomTagLevel.Instance, CustomTagStatus.Added);
+            CustomTagEntry formalized = _customTagEntryFormalizer.Formalize(entry);
+            Assert.Equal(tag.GetDefaultVR().Code, formalized.VR);
+        }
+
+        [Fact]
+        public void GivenStandardTagWithVR_WhenFormalizing_ThenVRShouldNotBeUpdated()
+        {
+            DicomTag tag = DicomTag.DeviceSerialNumber;
+            string vr = DicomVR.CS.Code;
+            CustomTagEntry entry = new CustomTagEntry(tag.GetPath(), vr, CustomTagLevel.Instance, CustomTagStatus.Added);
+            CustomTagEntry formalized = _customTagEntryFormalizer.Formalize(entry);
+            Assert.Equal(vr, formalized.VR);
+        }
+
+        [Fact]
+        public void GivenTagOfLowerCase_WhenFormalizing_ThenTagShouldBeUpperCase()
+        {
+            DicomTag tag = DicomTag.DeviceLabel;
+            CustomTagEntry entry = new CustomTagEntry(tag.GetPath().ToLowerInvariant(), tag.GetDefaultVR().Code, CustomTagLevel.Instance, CustomTagStatus.Added);
+            CustomTagEntry formalized = _customTagEntryFormalizer.Formalize(entry);
+            Assert.Equal(entry.Path.ToUpperInvariant(), formalized.Path);
+        }
+
+        [Fact]
+        public void GivenVROfLowerCase_WhenFormalizing_ThenVRShouldBeUpperCase()
+        {
+            DicomTag tag = DicomTag.DeviceLabel;
+            CustomTagEntry entry = new CustomTagEntry(tag.GetPath(), tag.GetDefaultVR().Code.ToLowerInvariant(), CustomTagLevel.Instance, CustomTagStatus.Added);
+            CustomTagEntry formalized = _customTagEntryFormalizer.Formalize(entry);
+            Assert.Equal(entry.VR.ToUpperInvariant(), formalized.VR);
+        }
+
+        [Fact]
+
+        public void GivenStandardTagAsKeyword_WhenFormalizing_ThenVRShouldBeFilled()
+        {
+            DicomTag tag = DicomTag.DeviceSerialNumber;
+            CustomTagEntry entry = new CustomTagEntry(path: tag.DictionaryEntry.Keyword, tag.GetDefaultVR().Code, CustomTagLevel.Instance, CustomTagStatus.Added);
+            string expectedPath = tag.GetPath();
+            CustomTagEntry formalized = _customTagEntryFormalizer.Formalize(entry);
+            Assert.Equal(formalized.Path, expectedPath);
+        }
+
+        public static IEnumerable<object[]> GetValidCustomTagEntries()
+        {
+            yield return new object[] { DicomTag.DeviceSerialNumber.BuildCustomTagEntry() }; // standard custom tag with VR
+            yield return new object[] { new CustomTagEntry("12051003", DicomVRCode.OB, CustomTagLevel.Instance, CustomTagStatus.Added) }; // private tag with VR
+        }
```

Comment:
I don't think you have to make a function for `MemberData`. Alternatively, and perhaps more simply, you can have a property or even field that is enumerable:

```csharp
public static IEnumerable<object[]> GetValidCustomTagEntries =>
    new object[][]
    {
        new object[] { DicomTag.DeviceSerialNumber.BuildCustomTagEntry() }, // standard custom tag with VR
        new object[] { new CustomTagEntry("12051003", DicomVRCode.OB, CustomTagLevel.Instance, CustomTagStatus.Added) }, // private tag with VR
    };
```

Reply:
This is really good idea, never think of this.